### PR TITLE
lib.trivial: rename NIX_ABORT_ON_WARN to NIXPKGS_ABORT_ON_WARN

### PR DESCRIPTION
**Pull request by stuebinm**
## Description of changes

Renames the `NIX_ABORT_ON_WARN` environment variable to `NIXPKGS_ABORT_ON_WARN`to make it consistent with other environment variables read by nixpkgs to configure its behaviour (e.g. `NIXPKGS_ALLOW_INSECURE` etc.). The previous naming appeared to suggest that the variable is read by Nix itself, which it is not. This is especially a source of confusion when used together with Nix's pure evaluation mode, in which case setting the variable has no effect.

The old `NIX_ABORT_ON_WARN` is still honoured, but produces a deprecation warning. This warning's check is added as a wrapper to the entire lib.trivial attrset, as it should be checked (and, if necessary, printed) regardless of whether or not a use of `lib.warn` is encountered, but whenever the `lib.warn` function is available (i.e. whenever `lib` itself is evaluated), so that users of `NIX_ABORT_ON_WARN` will see it even if their code does not currently produce any warnings. Likewise, this means that the warning does not actually abort evaluation, as otherwise downstream users of `NIX_ABORT_ON_WARN` would be immediately presented with an unrecoverable error.

This deprecation warning and honouring of the older behaviour can then be removed in a future release.

## Downsides

For reasons which remain mysterious to me, the deprecation warning is printed eight times if `nixpkgs` is imported as a flake:

~~~
$ NIX_ABORT_ON_WARN=1 nix build .#hello --impure
trace: warning: NIX_ABORT_ON_WARN has been renamed to NIXPKGS_ABORT_ON_WARN. NIX_ABORT_ON_WARN is still honored, but deprecated and will be removed from nixpkgs after the 24.11 release.
trace: warning: NIX_ABORT_ON_WARN has been renamed to NIXPKGS_ABORT_ON_WARN. NIX_ABORT_ON_WARN is still honored, but deprecated and will be removed from nixpkgs after the 24.11 release.
trace: warning: NIX_ABORT_ON_WARN has been renamed to NIXPKGS_ABORT_ON_WARN. NIX_ABORT_ON_WARN is still honored, but deprecated and will be removed from nixpkgs after the 24.11 release.
trace: warning: NIX_ABORT_ON_WARN has been renamed to NIXPKGS_ABORT_ON_WARN. NIX_ABORT_ON_WARN is still honored, but deprecated and will be removed from nixpkgs after the 24.11 release.
trace: warning: NIX_ABORT_ON_WARN has been renamed to NIXPKGS_ABORT_ON_WARN. NIX_ABORT_ON_WARN is still honored, but deprecated and will be removed from nixpkgs after the 24.11 release.
trace: warning: NIX_ABORT_ON_WARN has been renamed to NIXPKGS_ABORT_ON_WARN. NIX_ABORT_ON_WARN is still honored, but deprecated and will be removed from nixpkgs after the 24.11 release.
trace: warning: NIX_ABORT_ON_WARN has been renamed to NIXPKGS_ABORT_ON_WARN. NIX_ABORT_ON_WARN is still honored, but deprecated and will be removed from nixpkgs after the 24.11 release.
trace: warning: NIX_ABORT_ON_WARN has been renamed to NIXPKGS_ABORT_ON_WARN. NIX_ABORT_ON_WARN is still honored, but deprecated and will be removed from nixpkgs after the 24.11 release.
~~~

but only once if imported in the traditional way:

~~~
$ NIX_ABORT_ON_WARN=1 nix build -f . hello
trace: warning: NIX_ABORT_ON_WARN has been renamed to NIXPKGS_ABORT_ON_WARN. NIX_ABORT_ON_WARN is still honored, but deprecated and will be removed from nixpkgs after the 24.11 release.
~~~

This is only a minor annoyance, but does perhaps (?) point to an unrelated issue with how the nixpkgs flake handles importing `lib`.

## Alternatives

There are a few possible placements for the deprecation warning & its check in the code:
 - inside `lib.warn` itself: seems the most logical place to put it, but it would then only be shown if a warning is actually produced. As (presumably) most users of `NIX_ABORT_ON_WARN` will strive to keep their code free of warnings, this could lead to them never seeing the deprecation message.
 - outside of `lib/`, e.g. in `pkgs/top-level/impure.nix`: i am very hesitant to touch that file, since even slight mistakes would possibly have large consequences. However, it is the only place i could find where, if placed there, the warning is only ever printed once even when using nixpkgs as a flake.

## Question

Should this change be mentioned in the release notes? I was unable to find any clear guidelines on what (if any) changes to `lib` should get an entry there.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: lib/trivial.nix [hunk 1]
```diff
@@ -12,7 +12,20 @@ let
...
     warn;
-in {
+
+  # We have to use builtins here in place of lib.warn, as otherwise its
+  # definition would depend on itself. Further, we do not want this to
+  # actually abort even if NIX_ABORT_ON_WARN is set, as otherwise legacy
+  # uses of this variable would immediately result in an unrecoverable
+  # error where there should be a deprecation warning.
+  printDeprecationWarning =
+    if builtins.getEnv "NIX_ABORT_ON_WARN" != "" then
```
Comment (infinisil) [inline lib/trivial.nix:22]:
```suggestion
    if lib.isInOldestRelease 2405 && builtins.getEnv "NIX_ABORT_ON_WARN" != "" then
```

This needs some code rearranging to get working without infinite recursion, but this makes sure that there's a transition period where stable third party projects can switch to `NIXPKGS_ABORT_ON_WARN` without warnings for end-users.

File: lib/trivial.nix [hunk 2]
```diff
@@ -698,8 +711,9 @@ in {
     ```
   */
   warn =
-    if lib.elem (builtins.getEnv "NIX_ABORT_ON_WARN") ["1" "true" "yes"]
-    then msg: builtins.trace "[1;31mwarning: ${msg}[0m" (abort "NIX_ABORT_ON_WARN=true; warnings are treated as unrecoverable errors.")
+    if lib.elem (builtins.getEnv "NIXPKGS_ABORT_ON_WARN") [ "1" "true" "yes" ]
+    || lib.elem (builtins.getEnv "NIX_ABORT_ON_WARN") [ "1" "true" "yes" ]
```
Comment (infinisil) [inline lib/trivial.nix:715]: Let's have a comment here to mention that this one is deprecated, referring to `printDeprecationWarning`

**Review by infinisil — CHANGES_REQUESTED** [review]
Just minor thoughts, but overall this is looking good to me! This is a really tricky change, but your explanation makes a lot of sense!

The multiple messages for Flakes can be fixed by putting the `builtins.trace` before even the `{ lib }:` part of the file (`builtins.trace "..." ({ lib }: ...)`). But this does feel a bit too ugly..

So indeed I think it would be better to figure out why the Flake handling appears to import `lib` that many times. But this can be done in a separate PR :)

Ping @roberth as the one who introduced this environment variable

**Comment by roberth** [comment]
While your reasoning makes sense, my original goal was to build this functionality into Nix.
I haven't followed through on that, because at the time, there wasn't a Nix team yet, and it proved really hard to get anything merged.
So my preferred way forward is to still build this into Nix.

1. Add `builtins.warn`, which logs at the correct Nix log level instead of just ANSI codes at "info" level.
2. Add a Nix config option `abort-on-warn` that defaults to the `NIX_ABORT_ON_WARN` value.
3. Change `lib.warn` to `builtins.warn or ...`, where the current implementation will serve as a good polyfill.

This PR would not prevent that from happening, but it would be messier, as users first have to learn to use `NIXPKGS_ABORT_ON_WARN`, and then switch back to `NIX_ABORT_ON_WARN` or `--abort-on-warn`. If we leave Nixpkgs as is, there's no need to learn anything.

Also there's an argument to be made that `lib` is fairly universal, and users might not perceive themselves as interacting with Nixpkgs when they need this behavior. For instance, they might be trying to solve a problem in a module system application like home-manager, which is not Nixpkgs.

File: lib/trivial.nix (hunk 1 above)
Comment (stuebinm) [reply lib/trivial.nix]: I have done so, though aesthetically I don't particularly like the result. If you can think of any nicer way to do this, please let me know / feel free to edit this branch.

File: lib/trivial.nix (hunk 2 above)
Comment (stuebinm) [reply lib/trivial.nix:719]: I've added one

**Comment by lorenzleutgeb** [comment]
When I found out about this environment variable recently (https://github.com/ngi-nix/ngipkgs/issues/155 and consequently https://discourse.nixos.org/t/disallow-warnings-on-nix-flake-check/38964) I had this change in mind, too, but did not follow through. I support @stuebinm's reasoning. No matter what happens upstream in Nix, this environment variable should have the prefix `NIXPKGS_` not `NIX_` (as explained). If at some point a new built-in or similar appears, then there'll just have to be another change. I am one of the few (?) "users" of this environment variable and don't mind re-learning.

**Comment by roberth** [comment]
I've just opened https://github.com/NixOS/nix/pull/10592. Sorry for dropping the ball earlier.
I think the team can merge this after the release, if not sooner.

**Pull request closed without merging**